src/opt/Optimizer/TypedDesugarer.hs

```diff
@@ -0,0 +1,257 @@
+-- for x <- listA, y <- listB, z <- ListC do
+--      fun
+-- end
+--
+-- into listA.flatMap(listB.flatMap(listC.map(fun)))
+forDesugared :: Expr -> Expr
+forDesugared e@For{emeta, sources, body} =
+  let n = length sources
+      collectionType = getType $ collection $ head sources
+      callNameList = if (not (AST.AST.isCaptured e)) || (unitType == getType body) || (isRangeObjectType collectionType)
+                     then replicate n (Name "foreach")
+                     else replicate (n-1) (Name "flatMap") ++ [Name "map"]
+      revSources = reverse sources
+      elemType = getType body
+      desugaredFor = nestCalls emeta callNameList sources body elemType
+  in  desugaredFor
+forDesugared m = m
+
+nestCalls :: Meta.Meta Expr -> [Name] -> [ForSource] -> Expr -> Type -> Expr
+nestCalls meta (name:_) (fs:[]) body elemType = intoCall meta name fs body elemType
+nestCalls meta (name:restOfNames) (fs:restFS) body elemType =
+  let nestedCall = intoCall meta name fs body elemType
+  in nestCalls meta restOfNames restFS nestedCall elemType
+
+intoCall :: Meta.Meta Expr -> Name -> ForSource -> Expr -> Type -> Expr
+intoCall met callName ForSource{fsName, fsTy, collection} bodyOrMethodCall elemType =
+  if isRefType (getType collection)
+  then let param = [intoParam met Val fsName fsTy]
+           arguments = [intoClosure met param Nothing bodyOrMethodCall]
+           elemT = if callName == Name "foreach" -- this feels iffy
+                   then []
+                   else [elemType]
+       in  intoMethodCall met elemT collection callName arguments
+   else let param = [intoParam met Val fsName fsTy]
+            arguments = [intoClosure met param Nothing bodyOrMethodCall] ++ [collection]
+            elemT =  if callName == Name "foreach" -- this feels iffy
+                     then [fromMaybe intType fsTy]
+                     else [(fromMaybe intType fsTy), elemType]
+            name = intoQName callName
+       in   intoFunctionCall met elemT name arguments
+
+boxForInSeq :: Expr -> Expr
+boxForInSeq e@Seq{emeta, eseq} =
+  let newEseq = boxFor eseq []
+  in  e{eseq = newEseq}
+  where
+    boxFor :: [Expr] -> [Expr] -> [Expr]
+    boxFor [] newEseq =  newEseq
+    boxFor (ex:expr) newEseq
+      | isFor ex = newEseq ++ [forBoxed ex (boxFor expr [])]
+      | otherwise = boxFor expr (newEseq ++ [ex])
+    isFor For{} = True
+    isFor _ = False
+
+
+forBoxed :: Expr -> [Expr] -> Expr
+forBoxed for@For{emeta, sources, body} postForExpr =
+  let listOfVar = getVariables body
+      listOfVarNames = map getVarName listOfVar
+      getVarName VarAccess{qname} = qnlocal qname
+      unBoxed = unBox listOfVar
+      desugaredForWithFieldAccBody = forDesugared $ varBodyToFieldBody for [] listOfVarNames
+      bodyforLetBoxes = intoSeq emeta (desugaredForWithFieldAccBody:unBoxed ++ postForExpr)
+      boxLet = boxVar emeta listOfVar bodyforLetBoxes
+      newSeq = intoSeq emeta [boxLet]
+  in  newSeq
+
+getVariables :: Expr -> [Expr]
```

Comment:
my feeling is that if `getVariables` is only used in a function, maybe it makes more sense to create a locally scoped function, using `where`, where it is actually used.